src/crafting.cpp

```diff
@@ -469,37 +469,90 @@ static void finalize_crafted_item( item &newit )
     }
 }
 
-static void set_item_inventory( item &newit )
+static item *set_item_inventory( player &p, item &newit )
 {
+    item *ret_val = &null_item_reference();
     if( newit.made_of( LIQUID ) ) {
         g->handle_all_liquid( newit, PICKUP_RANGE );
     } else {
-        g->u.inv.assign_empty_invlet( newit, g->u );
+        p.inv.assign_empty_invlet( newit, p );
         // We might not have space for the item
-        if( !g->u.can_pickVolume( newit ) ) { //Accounts for result_mult
-            put_into_vehicle_or_drop( g->u, item_drop_reason::too_large, { newit } );
-        } else if( !g->u.can_pickWeight( newit, !get_option<bool>( "DANGEROUS_PICKUPS" ) ) ) {
-            put_into_vehicle_or_drop( g->u, item_drop_reason::too_heavy, { newit } );
+        if( !p.can_pickVolume( newit ) ) { //Accounts for result_mult
+            put_into_vehicle_or_drop( p, item_drop_reason::too_large, { newit } );
+        } else if( !p.can_pickWeight( newit, !get_option<bool>( "DANGEROUS_PICKUPS" ) ) ) {
+            put_into_vehicle_or_drop( p, item_drop_reason::too_heavy, { newit } );
         } else {
-            newit = g->u.i_add( newit );
-            add_msg( m_info, "%c - %s", newit.invlet == 0 ? ' ' : newit.invlet, newit.tname().c_str() );
+            ret_val = &p.i_add( newit );
+            add_msg( m_info, "%c - %s", ret_val->invlet == 0 ? ' ' : ret_val->invlet,
+                     ret_val->tname().c_str() );
         }
     }
+    return ret_val;
 }
 
-time_duration get_rot_since( const time_point &start, const time_point &end,
-                             const tripoint &location ); // weather.cpp
+static void return_all_components_for_craft( player &p, std::list<item> &used )
+{
+    // Force add here since the craft item is removed immediately after
+    for( item &it : used ) {
+        set_item_inventory( p, it );
+    }
+}
 
-void player::complete_craft()
+static void return_some_components_for_craft( player &p, std::list<item> &used )
+{
+    int count = 0;
+    for( item &it : used ) {
+        // Force add here since the craft item is removed in the same turn
+        // Each component has a 50% chance of being returned
+        // Never return the first component
+        if( count != 0 && one_in( 2 ) ) {
+            set_item_inventory( p, it );
+        }
+        ++count;
```

Comment:
It would be cleaner with the plain old iterators rather than with the range-based for loop.

Reply:
For sure, thanks!